File: src/parsers.ts
import { csv, dsv } from 'd3'
import { tail, head } from 'ramda'
import Axios from 'axios'

export interface Point { x: number, y: number, z: number }
interface RawSegment { 'Md': string, 'Inc': string, 'Azi': string }
export interface PathSegment { md: number, inc: number, azi: number, tvd: number, length: number }
export type Path = Array<PathSegment>
export type PipeSegment = { position: Point, rotation: Point, length: number, radius: number, md: number, tvd: number, pipeType: string }
export type PipeGeometry = Array<PipeSegment>
export interface GeometrySegment { name: string, radius: number, md: number }
export type GeometryDef = Array<GeometrySegment>

async function parsePipePressure() {
  return { mds: require('./data/mds.json') }
}

async function parseWellPath(): Promise<Path> {
  const path = await dsv(';', require('./data/well_path.csv')) as Array<RawSegment>
  const result = tail(
    path.map(segment => ({
      md: Number(segment['Md']),
      tvd: Number(segment['Tvd']),
      inc: Number(segment['Inc']) * Math.PI / 180,
      azi: Number(segment['Azi']) * Math.PI / 180,
    })).map((segment, i, arr) => ((i === 0) ? null : {
      length: arr[i].md - arr[i-1].md,
      ...segment
    }))
  )
  return result
}

async function parseGeometrydef() {
  const response = await Axios.get(require('./data/geometrydef.txt'))
  
  const data = response.data.split('\n')
    .map((line: string) => head(line.split('#'))) // Remove comments
    .map(Number)

  const inchesToMetres = (inches: number) => 0.0254 * inches

  return [
    { name: 'riser',         radius: inchesToMetres(data[1]) / 2, md: data[2]           },
    { name: 'cased section', radius: inchesToMetres(data[3]) / 2, md: data[4]           },
    { name: 'liner',         radius: inchesToMetres(data[5]) / 2, md: data[4] + data[6] },
    { name: 'open hole',     radius: inchesToMetres(data[0]) / 2, md: data[10]          }
  ]
}

export const allDataPromise = Promise.all([
  parseWellPath(),
  parseGeometrydef(),
  parsePipePressure()
])

File: services/api/app/parsers.py
import pandas as pd
import numpy as np
from math import pi

meters_per_inch = 0.0254

def _to_float_or_none(n: str):
    try:
        return float(n.strip())
    except:
        return None



def parse_pipepressure(path: str):
    return pd.read_csv(f'{path}/pipepressure.csv', index_col=0).rename(columns=_to_float_or_none)


def parse_pipestress(path: str):
    return pd.read_csv(f'{path}/pipepressure.csv', index_col=0).rename(columns=_to_float_or_none)


def parse_well_path(path: str):
    return pd.read_csv(f'{path}/well_path.csv', sep=';')\
        .rename(columns={ 'Md': 'md', 'Inc': 'rotz', 'Azi': 'roty', 'Tvd': 'tvd', 'TVD(m)': 'tvd' })\
        [['md', 'roty', 'rotz', 'tvd']]


def parse_mud_density(path: str):
    with open(f'{path}/fluiddef.txt') as fp:
        first_line = fp.readline()
        return float(first_line.split('#')[0].strip())


def parse_geometrydef(path: str):
    with open(f'{path}/geometrydef.txt') as fp:
        all_lines = fp.readlines()
        data = map(lambda line: line.split('#'), all_lines)    # Split up numbers and descriptions
        data = [item for sublist in data for item in sublist]  # Flatten the results
        data = map(_to_float_or_none, data)
        data = filter(lambda number: number is not None, data)
        data = list(data)

        radii = list(map(
            lambda inches: 0.5 * meters_per_inch * inches, 
            [ data[1], data[3], data[5], data[0] ]
        ))
        mds = [ 0, data[2], data[4], data[4] + data[6], data[10] ]

        return list(filter(lambda x: x['length'] > 0, [
            { 'name': 'riser',         'radius': radii[0], 'md_start': mds[0], 'md_stop': mds[1], 'length': mds[1] - mds[0] },
            { 'name': 'cased section', 'radius': radii[1], 'md_start': mds[1], 'md_stop': mds[2], 'length': mds[2] - mds[1] },
            { 'name': 'liner',         'radius': radii[2], 'md_start': mds[2], 'md_stop': mds[3], 'length': mds[3] - mds[2] },
            { 'name': 'open hole',     'radius': radii[3], 'md_start': mds[3], 'md_stop': mds[4], 'length': mds[4] - mds[3] }
        ]))


